File: modules/libs/wire/agent.ts
/**
 * The agent port, answered from the wire.
 *
 * Every window asks the same agent the same way, so what a step on the wire
 * comes to in the port is written down once. The window makes the client; this
 * says what its answers mean.
 */
import type { AskAgentResponse, Span } from '@numen/protocol'
import type { AgentPort } from '@numen/ui'

/** The span a call named, and one naming no place inside the source. */
const spanOf = (span: Span | undefined) => ({ from: span?.from ?? 0, to: span?.to ?? 0 })

/** As much of the agent service as the port asks of it. */
export interface AgentClient {
  askAgent(
    request: { asked: string; focus: string; conversation: string },
    options: { signal: AbortSignal },
  ): AsyncIterable<AskAgentResponse>
  finishConversation(request: { conversation: string }): Promise<unknown>
}

/** The agent port, over the client the window talks on. */
export const agentPort = (agent: AgentClient): AgentPort => ({
  async *ask(asked, focus, conversation, signal) {
    for await (const step of agent.askAgent({ asked, focus, conversation }, { signal })) {
      switch (step.step.case) {
        case 'said':
          yield { kind: 'said', text: step.step.value }
          break
        case 'toolCall': {
          const said = step.step.value
          yield {
            kind: 'toolCall',
            tool: said.tool,
            about: said.about,
            written: said.written,
            ...(said.path ? { place: { path: said.path, span: spanOf(said.span) } } : {}),
          }
          break
        }
        case 'answered':
          yield { kind: 'answered' }
          break
        case 'thinking':
          yield { kind: 'thinking' }
          break
        case 'stopped':
          yield { kind: 'stopped', error: step.step.value }
          break
      }
    }
  },

  async finish(conversation) {
    await agent.finishConversation({ conversation })
  },
})

File: modules/libs/core/internal/wire/agent.go
package wire

import (
	"context"
	"time"

	"connectrpc.com/connect"

	v1 "github.com/jiva-studio/numen/modules/libs/protocol/gen/numen/v1"

	"github.com/jiva-studio/numen/modules/libs/core/domain"
	"github.com/jiva-studio/numen/modules/libs/core/port"
)

// Ask hands one task to the agent and writes what it does to the stream, for as
// long as the client listens.
//
// The work is stopped on the way out, whether it finished, failed or the client
// went away. Nothing outlives the panel it was asked from.
//
// A model thinking for a minute writes nothing, and a stream that writes
// nothing never learns its client has gone. A step naming nothing is one a
// client ignores and a write that fails ends the work.
func Ask(
	ctx context.Context,
	taking port.Agent,
	task port.Task,
	stream *connect.ServerStream[v1.AskAgentResponse],
) error {
	work, err := taking.Take(ctx, task)
	if err != nil {
		return connect.NewError(connect.CodeInternal, err)
	}
	defer work.Stop()

	repeat := time.NewTicker(Again)
	defer repeat.Stop()

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-repeat.C:
			if err := stream.Send(&v1.AskAgentResponse{}); err != nil {
				return err
			}
		case step, working := <-work.Steps():
			if !working {
				return nil
			}
			for _, out := range stepsOf(step) {
				if err := stream.Send(out); err != nil {
					return err
				}
			}
			if step.Kind == port.StepStopped {
				return nil
			}
		}
	}
}

// stepsOf says a step of an agent's work in the schema's words.
//
// Every kind that names a tool is drawn as one, whatever that tool does to the
// vault, and carries where in the vault it is working.
func stepsOf(step port.Step) []*v1.AskAgentResponse {
	switch step.Kind {
	case port.StepToolCall, port.StepRead, port.StepEdit,
		port.StepRemove, port.StepMove, port.StepSearch:
		return []*v1.AskAgentResponse{{Step: &v1.AskAgentResponse_ToolCall{
			ToolCall: &v1.ToolCall{
				Tool:    step.Tool,
				About:   step.About,
				Written: int32(step.Count),
				Path:    step.Place.Path,
				Span:    spanOf(step.Place.Spans),
			},
		}}}
	case port.StepAnswered:
		return []*v1.AskAgentResponse{{Step: &v1.AskAgentResponse_Answered{Answered: &v1.Answered{}}}}
	case port.StepThinking:
		return []*v1.AskAgentResponse{{Step: &v1.AskAgentResponse_Thinking{Thinking: &v1.Thinking{}}}}
	case port.StepStopped:
		return []*v1.AskAgentResponse{{Step: &v1.AskAgentResponse_Stopped{Stopped: step.Detail}}}
	default:
		return []*v1.AskAgentResponse{{Step: &v1.AskAgentResponse_Said{Said: step.Text}}}
	}
}

// spanOf is the one span a call names, and nothing where it named the source
// and no place inside it.
func spanOf(spans []domain.ByteSpan) *v1.Span {
	if len(spans) == 0 {
		return nil
	}
	return &v1.Span{From: int32(spans[0].From), To: int32(spans[0].To)}
}
